src/peft/tuners/lora/layer.py

```diff
@@ -149,6 +152,45 @@ def reset_lora_parameters(self, adapter_name, init_lora_weights):
             nn.init.zeros_(self.lora_embedding_A[adapter_name])
             nn.init.normal_(self.lora_embedding_B[adapter_name])
 
+    def pissa_init(self, adapter_name, init_lora_weights):
+        assert self.scaling[adapter_name] == 1
+        weight = self.get_base_layer().weight
+        dtype = weight.dtype
+        device = weight.device
+        if dtype == torch.uint8:
+            quant_type = weight.quant_type
+            import bitsandbytes as bnb
+
+            weight = bnb.functional.dequantize_4bit(weight.data, weight.quant_state).to(torch.float32)
+        elif dtype != torch.float32:
+            weight = self.get_base_layer().weight.to(torch.float32)
+
+        if init_lora_weights == "pissa":
+            U, S, Vh = torch.linalg.svd(weight.data, full_matrices=False)
+            Ur = U[:, : self.r[adapter_name]]
+            Sr = S[: self.r[adapter_name]]
+            Vhr = Vh[: self.r[adapter_name]]
+        elif len(init_lora_weights.split("_niter_")) == 2:
+            Ur, Sr, Vr = svd_lowrank(
+                weight.data, self.r[adapter_name], niter=int(init_lora_weights.split("_niter_")[-1])
+            )
+            Vhr = Vr.t()
+        else:
+            raise "init_lora_weights should be pissa or pissa_niter_[number of iters]."
+
+        lora_A = torch.diag(torch.sqrt(Sr)) @ Vhr
+        lora_B = Ur @ torch.diag(torch.sqrt(Sr))
+        self.lora_A[adapter_name].weight.data = lora_A
+        self.lora_B[adapter_name].weight.data = lora_B
+        res = weight.data - lora_B @ lora_A
+        if dtype == torch.uint8:
+            weight = bnb.nn.Params4bit(
+                res.to("cpu"), requires_grad=False, compress_statistics=False, quant_type=quant_type
+            ).to(device)
+        else:
+            weight = res.to(dtype)
+        self.get_base_layer().weight.data = weight
```

Comment:
If I'm not mistaken, this part is a bit problematic. Let me explain based on an example: Say a user wants to fine-tune Llama 7b. They will load this model with:

```python
base_model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b")  # <= loads original Llama 2 weights
lora_config = LoraConfig(..., init_lora_weights="pissa")
peft_model = get_peft_model(base_model, lora_config)  # <= this mutates the base weights of Llama 2

# training code ...

peft_model.save_pretrained(<PATH>)
```

Later, when they load their code:

```python
base_model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b")  # <= loads original Llama 2 weights
peft_model = PeftModel.from_pretrained(base_model, <PATH>)
```

Now, the loaded LoRA adapters do _not_ fit the weights for the base model and we will get different results.

What can we do? Here are some possibilities:

1. We have to save the mutated base weights as well. However, this is very expensive, since we need to save a full copy of the weights. This negates one advantage of LoRA, i.e. that this has a very small file size.
2. Can we skip mutating the base weights? How much performance would we lose if we only use the new LoRA A and B weights and keep the base weights the same?
3. We have to run the same mutation steps of the base weights when we load the PiSSA weights. This might be the most pragmatic solution from the user's point of view, but it will be very hard (or even impossible?) to implement.